File: multi_car_diary.cs
// multi_car_diary.cs — C# версия

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;

class Repair
{
    public string Date { get; set; }
    public int Mileage { get; set; }
    public string Description { get; set; }
    public double Cost { get; set; }
}

class OilChange
{
    public string Date { get; set; }
    public int Mileage { get; set; }
    public string OilType { get; set; }
    public string Filter { get; set; }
    public double Cost { get; set; }
    public int Interval { get; set; }
    public int NextMileage { get; set; }
}

class Reminder
{
    public string Title { get; set; }
    public string Date { get; set; }
    public int Mileage { get; set; }
    public string Description { get; set; }
    public string Priority { get; set; }
    public bool Done { get; set; }
}

class Car
{
    public int Id { get; set; }
    public string Brand { get; set; }
    public string Model { get; set; }
    public int Year { get; set; }
    public string Vin { get; set; }
    public int Mileage { get; set; }
    public List<Repair> Repairs { get; set; } = new List<Repair>();
    public List<OilChange> OilChanges { get; set; } = new List<OilChange>();
    public List<Reminder> Reminders { get; set; } = new List<Reminder>();
}

class Program
{
    private static List<Car> cars = new List<Car>();
    private const string DataFile = "cars.json";

    static void Main()
    {
        Load();
        while (true)
        {
            Console.WriteLine("\n\u001B[36m🚗 Многомарочный дневник автомобиля (C#)\u001B[0m");
            Console.WriteLine("1. Выбрать/создать автомобиль");
            Console.WriteLine("2. Удалить автомобиль");
            Console.WriteLine("3. Показать все автомобили");
            Console.WriteLine("4. Выход");
            Console.Write("Выберите действие: ");
            string choice = Console.ReadLine();
            switch (choice)
            {
                case "1": SelectOrCreateCar(); break;
                case "2": DeleteCar(); break;
                case "3": ListCars(); break;
                case "4": Console.WriteLine("До свидания!"); return;
                default: Console.WriteLine("\u001B[31mНеверный выбор.\u001B[0m"); break;
            }
        }
    }

    static void Load()
    {
        if (File.Exists(DataFile))
        {
            try
            {
                string json = File.ReadAllText(DataFile);
                cars = JsonSerializer.Deserialize<List<Car>>(json) ?? new List<Car>();
            }
            catch { cars = new List<Car>(); }
        }
    }

    static void Save()
    {
        string json = JsonSerializer.Serialize(cars, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(DataFile, json);
    }

    static Car GetCar(int id) => cars.FirstOrDefault(c => c.Id == id);

    static void ListCars()
    {
        if (cars.Count == 0)
        {
            Console.WriteLine("\u001B[33mНет автомобилей.\u001B[0m");
            return;
        }
        Console.WriteLine($"\u001B[36m{"ID",-4} {"Марка",-15} {"Модель",-15} {"Год",-6} {"Пробег",-10}\u001B[0m");
        Console.WriteLine(new string('-', 55));
        foreach (var c in cars)
            Console.WriteLine($"{c.Id,-4} {c.Brand,-15} {c.Model,-15} {c.Year,-6} {c.Mileage,-10}");
    }

    static void SelectOrCreateCar()
    {
        ListCars();
        if (cars.Count > 0)
        {
            Console.Write("Введите номер автомобиля (или 0 для создания нового): ");
            string input = Console.ReadLine();
            if (input == "0")
            {
                CreateNewCar();
            }
            else if (int.TryParse(input, out int id))
            {
                var car = GetCar(id);
                if (car != null)
                    CarMenu(car);
                else
                    Console.WriteLine("\u001B[31m❌ Автомобиль не найден.\u001B[0m");
            }
        }
        else
        {
            CreateNewCar();
        }
    }

    static void CreateNewCar()
    {
        Console.Write("Марка: ");
        string brand = Console.ReadLine();
        Console.Write("Модель: ");
        string model = Console.ReadLine();
        Console.Write("Год: ");
        int year = int.Parse(Console.ReadLine());
        Console.Write("VIN: ");
        string vin = Console.ReadLine();
        Console.Write("Текущий пробег (км): ");
        int mileage = int.Parse(Console.ReadLine());
        int id = cars.Count + 1;
        cars.Add(new Car { Id = id, Brand = brand, Model = model, Year = year, Vin = vin, Mileage = mileage });
        Save();
        Console.WriteLine($"\u001B[32m✅ Автомобиль добавлен (ID: {id})\u001B[0m");
        var car = GetCar(id);
        if (car != null) CarMenu(car);
    }

    static void DeleteCar()
    {
        ListCars();
        Console.Write("Введите ID автомобиля для удаления: ");
        int id = int.Parse(Console.ReadLine());
        var car = GetCar(id);
        if (car != null)
        {
            cars.Remove(car);
            Save();
            Console.WriteLine("\u001B[32m✅ Автомобиль удалён.\u001B[0m");
        }
        else
            Console.WriteLine("\u001B[31m❌ Автомобиль не найден.\u001B[0m");
    }

    static void CarMenu(Car car)
    {
        while (true)
        {
            Console.WriteLine($"\n\u001B[36m🚗 {car.Brand} {car.Model} {car.Year} (Пробег: {car.Mileage} км) — меню автомобиля\u001B[0m");
            Console.WriteLine("1. Добавить ремонт");
            Console.WriteLine("2. Добавить замену масла");
            Console.WriteLine("3. Добавить напоминание");
            Console.WriteLine("4. Показать историю");
            Console.WriteLine("5. Показать статистику");
            Console.WriteLine("6. Назад");
            Console.Write("Выберите действие: ");
            string choice = Console.ReadLine();
            switch (choice)
            {
                case "1": AddRepair(car); break;
                case "2": AddOilChange(car); break;
                case "3": AddReminder(car); break;
                case "4": ShowHistory(car); break;
                case "5": ShowStats(car); break;
                case "6": return;
                default: Console.WriteLine("\u001B[31mНеверный выбор.\u001B[0m"); break;
            }
        }
    }

    static void AddRepair(Car car)
    {
        Console.Write("Дата (ГГГГ-ММ-ДД): ");
        string date = Console.ReadLine();
        Console.Write("Пробег (км): ");
        int mileage = int.Parse(Console.ReadLine());
        Console.Write("Описание: ");
        string desc = Console.ReadLine();
        Console.Write("Стоимость (руб): ");
        double cost = double.Parse(Console.ReadLine());
        car.Repairs.Add(new Repair { Date = date, Mileage = mileage, Description = desc, Cost = cost });
        if (mileage > car.Mileage) car.Mileage = mileage;
        Save();
        Console.WriteLine("\u001B[32m✅ Ремонт добавлен.\u001B[0m");
    }

    static void AddOilChange(Car car)
    {
        Console.Write("Дата (ГГГГ-ММ-ДД): ");
        string date = Console.ReadLine();
        Console.Write("Пробег (км): ");
        int mileage = int.Parse(Console.ReadLine());
        Console.Write("Тип масла: ");
        string oil = Console.ReadLine();
        Console.Write("Фильтр (артикул): ");
        string filter = Console.ReadLine();
        Console.Write("Стоимость (руб): ");
        double cost = double.Parse(Console.ReadLine());
        Console.Write("Интервал (км, по умолч. 10000): ");
        string intervalStr = Console.ReadLine();
        int interval = string.IsNullOrEmpty(intervalStr) ? 10000 : int.Parse(intervalStr);
        car.OilChanges.Add(new OilChange { Date = date, Mileage = mileage, OilType = oil, Filter = filter, Cost = cost, Interval = interval, NextMileage = mileage + interval });
        if (mileage > car.Mileage) car.Mileage = mileage;
        Save();
        Console.WriteLine("\u001B[32m✅ Замена масла добавлена.\u001B[0m");
    }

    static void AddReminder(Car car)
    {
        Console.Write("Название: ");
        string title = Console.ReadLine();
        Console.Write("Дата (ГГГГ-ММ-ДД): ");
        string date = Console.ReadLine();
        Console.Write("Пробег (км): ");
        int mileage = int.Parse(Console.ReadLine());
        Console.Write("Описание: ");
        string desc = Console.ReadLine();
        Console.Write("Приоритет (низкий/средний/высокий): ");
        string priority = Console.ReadLine().ToLower();
        if (priority != "низкий" && priority != "средний" && priority != "высокий") priority = "средний";
        car.Reminders.Add(new Reminder { Title = title, Date = date, Mileage = mileage, Description = desc, Priority = priority, Done = false });
        Save();
        Console.WriteLine("\u001B[32m✅ Напоминание добавлено.\u001B[0m");
    }

    static void ShowHistory(Car car)
    {
        Console.WriteLine($"\n\u001B[36m📋 История для {car.Brand} {car.Model} {car.Year}\u001B[0m");
        if (car.Repairs.Count > 0)
        {
            Console.WriteLine("\u001B[33mРемонты:\u001B[0m");
            foreach (var r in car.Repairs)
                Console.WriteLine($"  {r.Date} | {r.Mileage} км | {r.Description} | {r.Cost} руб.");
        }
        if (car.OilChanges.Count > 0)
        {
            Console.WriteLine("\u001B[33mЗамены масла:\u001B[0m");
            foreach (var o in car.OilChanges)
                Console.WriteLine($"  {o.Date} | {o.Mileage} км | {o.OilType} | {o.Filter} | {o.Cost} руб. | след. {o.NextMileage} км");
        }
        if (car.Reminders.Count > 0)
        {
            Console.WriteLine("\u001B[33mНапоминания:\u001B[0m");
            foreach (var r in car.Reminders)
                Console.WriteLine($"  {(r.Done ? "✅" : "⏳")} {r.Title} | {r.Date} | {r.Priority} | {r.Description}");
        }
    }

    static void ShowStats(Car car)
    {
        double totalRepair = car.Repairs.Sum(r => r.Cost);
        double totalOil = car.OilChanges.Sum(o => o.Cost);
        double total = totalRepair + totalOil;
        Console.WriteLine($"\n\u001B[36m📊 Статистика для {car.Brand} {car.Model} {car.Year}\u001B[0m");
        Console.WriteLine($"  Всего ремонтов: {car.Repairs.Count}");
        Console.WriteLine($"  Всего замен масла: {car.OilChanges.Count}");
        Console.WriteLine($"  Всего напоминаний: {car.Reminders.Count}");
        Console.WriteLine($"  Общая стоимость: {total:F2} руб.");
        if (car.Repairs.Count > 0)
            Console.WriteLine($"  Средняя стоимость ремонта: {totalRepair / car.Repairs.Count:F2} руб.");
    }
}

File: multi_car_diary.java
// multi_car_diary.java — Java версия

import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.time.*;
import java.time.format.*;

class Repair {
    String date;
    int mileage;
    String description;
    double cost;

    Repair(String date, int mileage, String description, double cost) {
        this.date = date;
        this.mileage = mileage;
        this.description = description;
        this.cost = cost;
    }

    String toJson() {
        return String.format("{\"date\":\"%s\",\"mileage\":%d,\"description\":\"%s\",\"cost\":%.2f}",
                date, mileage, description, cost);
    }
}

class OilChange {
    String date;
    int mileage;
    String oilType;
    String filter;
    double cost;
    int interval;
    int nextMileage;

    OilChange(String date, int mileage, String oilType, String filter, double cost, int interval, int nextMileage) {
        this.date = date;
        this.mileage = mileage;
        this.oilType = oilType;
        this.filter = filter;
        this.cost = cost;
        this.interval = interval;
        this.nextMileage = nextMileage;
    }

    String toJson() {
        return String.format("{\"date\":\"%s\",\"mileage\":%d,\"oil_type\":\"%s\",\"filter\":\"%s\",\"cost\":%.2f,\"interval\":%d,\"next_mileage\":%d}",
                date, mileage, oilType, filter, cost, interval, nextMileage);
    }
}

class Reminder {
    String title;
    String date;
    int mileage;
    String description;
    String priority;
    boolean done;

    Reminder(String title, String date, int mileage, String description, String priority, boolean done) {
        this.title = title;
        this.date = date;
        this.mileage = mileage;
        this.description = description;
        this.priority = priority;
        this.done = done;
    }

    String toJson() {
        return String.format("{\"title\":\"%s\",\"date\":\"%s\",\"mileage\":%d,\"description\":\"%s\",\"priority\":\"%s\",\"done\":%b}",
                title, date, mileage, description, priority, done);
    }
}

class Car {
    int id;
    String brand;
    String model;
    int year;
    String vin;
    int mileage;
    List<Repair> repairs;
    List<OilChange> oilChanges;
    List<Reminder> reminders;

    Car(int id, String brand, String model, int year, String vin, int mileage) {
        this.id = id;
        this.brand = brand;
        this.model = model;
        this.year = year;
        this.vin = vin;
        this.mileage = mileage;
        this.repairs = new ArrayList<>();
        this.oilChanges = new ArrayList<>();
        this.reminders = new ArrayList<>();
    }

    String toJson() {
        StringBuilder sb = new StringBuilder();
        sb.append("{\"id\":").append(id).append(",\"brand\":\"").append(brand).append("\",\"model\":\"").append(model)
          .append("\",\"year\":").append(year).append(",\"vin\":\"").append(vin).append("\",\"mileage\":").append(mileage)
          .append(",\"repairs\":[");
        for (int i = 0; i < repairs.size(); i++) {
            sb.append(repairs.get(i).toJson());
            if (i < repairs.size() - 1) sb.append(",");
        }
        sb.append("],\"oil_changes\":[");
        for (int i = 0; i < oilChanges.size(); i++) {
            sb.append(oilChanges.get(i).toJson());
            if (i < oilChanges.size() - 1) sb.append(",");
        }
        sb.append("],\"reminders\":[");
        for (int i = 0; i < reminders.size(); i++) {
            sb.append(reminders.get(i).toJson());
            if (i < reminders.size() - 1) sb.append(",");
        }
        sb.append("]}");
        return sb.toString();
    }
}

public class multi_car_diary {
    private static final String DATA_FILE = "cars.json";
    private static List<Car> cars = new ArrayList<>();
    private static Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) {
        load();
        while (true) {
            System.out.println("\n\u001B[36m🚗 Многомарочный дневник автомобиля (Java)\u001B[0m");
            System.out.println("1. Выбрать/создать автомобиль");
            System.out.println("2. Удалить автомобиль");
            System.out.println("3. Показать все автомобили");
            System.out.println("4. Выход");
            System.out.print("Выберите действие: ");
            String choice = scanner.nextLine();
            switch (choice) {
                case "1": selectOrCreateCar(); break;
                case "2": deleteCar(); break;
                case "3": listCars(); break;
                case "4": System.out.println("До свидания!"); return;
                default: System.out.println("\u001B[31mНеверный выбор.\u001B[0m");
            }
        }
    }

    private static void load() {
        try {
            String content = new String(Files.readAllBytes(Paths.get(DATA_FILE)));
            // Упрощённый парсинг, в реальности нужен JSON-парсер
            cars = new ArrayList<>();
        } catch (IOException e) {
            cars = new ArrayList<>();
        }
    }

    private static void save() {
        try {
            StringBuilder sb = new StringBuilder("[");
            for (int i = 0; i < cars.size(); i++) {
                sb.append(cars.get(i).toJson());
                if (i < cars.size() - 1) sb.append(",");
            }
            sb.append("]");
            Files.write(Paths.get(DATA_FILE), sb.toString().getBytes());
        } catch (IOException e) {
            System.out.println("Ошибка сохранения.");
        }
    }

    private static void listCars() {
        if (cars.isEmpty()) {
            System.out.println("\u001B[33mНет автомобилей.\u001B[0m");
            return;
        }
        System.out.printf("\u001B[36m%-4s %-15s %-15s %-6s %-10s\u001B[0m\n", "ID", "Марка", "Модель", "Год", "Пробег");
        System.out.println("-".repeat(55));
        for (Car c : cars) {
            System.out.printf("%-4d %-15s %-15s %-6d %-10d\n", c.id, c.brand, c.model, c.year, c.mileage);
        }
    }

    private static Car getCar(int id) {
        for (Car c : cars) {
            if (c.id == id) return c;
        }
        return null;
    }

    private static int addCar(String brand, String model, int year, String vin, int mileage) {
        int id = cars.size() + 1;
        cars.add(new Car(id, brand, model, year, vin, mileage));
        save();
        return id;
    }

    private static void deleteCar() {
        listCars();
        System.out.print("Введите ID автомобиля для удаления: ");
        int id = Integer.parseInt(scanner.nextLine());
        boolean removed = cars.removeIf(c -> c.id == id);
        if (removed) {
            save();
            System.out.println("\u001B[32m✅ Автомобиль удалён.\u001B[0m");
        } else {
            System.out.println("\u001B[31m❌ Автомобиль не найден.\u001B[0m");
        }
    }

    private static void selectOrCreateCar() {
        listCars();
        if (!cars.isEmpty()) {
            System.out.print("Введите номер автомобиля (или 0 для создания нового): ");
            String input = scanner.nextLine();
            if (input.equals("0")) {
                createNewCar();
            } else {
                int id = Integer.parseInt(input);
                Car car = getCar(id);
                if (car != null) {
                    carMenu(car);
                } else {
                    System.out.println("\u001B[31m❌ Автомобиль не найден.\u001B[0m");
                }
            }
        } else {
            createNewCar();
        }
    }

    private static void createNewCar() {
        System.out.print("Марка: ");
        String brand = scanner.nextLine();
        System.out.print("Модель: ");
        String model = scanner.nextLine();
        System.out.print("Год: ");
        int year = Integer.parseInt(scanner.nextLine());
        System.out.print("VIN: ");
        String vin = scanner.nextLine();
        System.out.print("Текущий пробег (км): ");
        int mileage = Integer.parseInt(scanner.nextLine());
        int id = addCar(brand, model, year, vin, mileage);
        System.out.println("\u001B[32m✅ Автомобиль добавлен (ID: " + id + ")\u001B[0m");
        Car car = getCar(id);
        if (car != null) carMenu(car);
    }

    private static void carMenu(Car car) {
        while (true) {
            System.out.println("\n\u001B[36m🚗 " + car.brand + " " + car.model + " " + car.year + " (Пробег: " + car.mileage + " км) — меню автомобиля\u001B[0m");
            System.out.println("1. Добавить ремонт");
            System.out.println("2. Добавить замену масла");
            System.out.println("3. Добавить напоминание");
            System.out.println("4. Показать историю");
            System.out.println("5. Показать статистику");
            System.out.println("6. Назад");
            System.out.print("Выберите действие: ");
            String choice = scanner.nextLine();
            switch (choice) {
                case "1": addRepair(car); break;
                case "2": addOilChange(car); break;
                case "3": addReminder(car); break;
                case "4": showHistory(car); break;
                case "5": showStats(car); break;
                case "6": return;
                default: System.out.println("\u001B[31mНеверный выбор.\u001B[0m");
            }
        }
    }

    private static void addRepair(Car car) {
        System.out.print("Дата (ГГГГ-ММ-ДД): ");
        String date = scanner.nextLine();
        System.out.print("Пробег (км): ");
        int mileage = Integer.parseInt(scanner.nextLine());
        System.out.print("Описание: ");
        String desc = scanner.nextLine();
        System.out.print("Стоимость (руб): ");
        double cost = Double.parseDouble(scanner.nextLine());
        car.repairs.add(new Repair(date, mileage, desc, cost));
        if (mileage > car.mileage) car.mileage = mileage;
        save();
        System.out.println("\u001B[32m✅ Ремонт добавлен.\u001B[0m");
    }

    private static void addOilChange(Car car) {
        System.out.print("Дата (ГГГГ-ММ-ДД): ");
        String date = scanner.nextLine();
        System.out.print("Пробег (км): ");
        int mileage = Integer.parseInt(scanner.nextLine());
        System.out.print("Тип масла: ");
        String oil = scanner.nextLine();
        System.out.print("Фильтр (артикул): ");
        String filter = scanner.nextLine();
        System.out.print("Стоимость (руб): ");
        double cost = Double.parseDouble(scanner.nextLine());
        System.out.print("Интервал (км, по умолч. 10000): ");
        String intervalStr = scanner.nextLine();
        int interval = intervalStr.isEmpty() ? 10000 : Integer.parseInt(intervalStr);
        car.oilChanges.add(new OilChange(date, mileage, oil, filter, cost, interval, mileage + interval));
        if (mileage > car.mileage) car.mileage = mileage;
        save();
        System.out.println("\u001B[32m✅ Замена масла добавлена.\u001B[0m");
    }

    private static void addReminder(Car car) {
        System.out.print("Название: ");
        String title = scanner.nextLine();
        System.out.print("Дата (ГГГГ-ММ-ДД): ");
        String date = scanner.nextLine();
        System.out.print("Пробег (км): ");
        int mileage = Integer.parseInt(scanner.nextLine());
        System.out.print("Описание: ");
        String desc = scanner.nextLine();
        System.out.print("Приоритет (низкий/средний/высокий): ");
        String priority = scanner.nextLine().toLowerCase();
        if (!priority.equals("низкий") && !priority.equals("средний") && !priority.equals("высокий")) {
            priority = "средний";
        }
        car.reminders.add(new Reminder(title, date, mileage, desc, priority, false));
        save();
        System.out.println("\u001B[32m✅ Напоминание добавлено.\u001B[0m");
    }

    private static void showHistory(Car car) {
        System.out.println("\n\u001B[36m📋 История для " + car.brand + " " + car.model + " " + car.year + "\u001B[0m");
        if (!car.repairs.isEmpty()) {
            System.out.println("\u001B[33mРемонты:\u001B[0m");
            for (Repair r : car.repairs) {
                System.out.printf("  %s | %d км | %s | %.2f руб.\n", r.date, r.mileage, r.description, r.cost);
            }
        }
        if (!car.oilChanges.isEmpty()) {
            System.out.println("\u001B[33mЗамены масла:\u001B[0m");
            for (OilChange o : car.oilChanges) {
                System.out.printf("  %s | %d км | %s | %s | %.2f руб. | след. %d км\n", o.date, o.mileage, o.oilType, o.filter, o.cost, o.nextMileage);
            }
        }
        if (!car.reminders.isEmpty()) {
            System.out.println("\u001B[33mНапоминания:\u001B[0m");
            for (Reminder r : car.reminders) {
                String status = r.done ? "✅" : "⏳";
                System.out.printf("  %s %s | %s | %s | %s\n", status, r.title, r.date, r.priority, r.description);
            }
        }
    }

    private static void showStats(Car car) {
        double totalRepair = car.repairs.stream().mapToDouble(r -> r.cost).sum();
        double totalOil = car.oilChanges.stream().mapToDouble(o -> o.cost).sum();
        double total = totalRepair + totalOil;
        System.out.println("\n\u001B[36m📊 Статистика для " + car.brand + " " + car.model + " " + car.year + "\u001B[0m");
        System.out.println("  Всего ремонтов: " + car.repairs.size());
        System.out.println("  Всего замен масла: " + car.oilChanges.size());
        System.out.println("  Всего напоминаний: " + car.reminders.size());
        System.out.printf("  Общая стоимость: %.2f руб.\n", total);
        if (!car.repairs.isEmpty()) {
            System.out.printf("  Средняя стоимость ремонта: %.2f руб.\n", totalRepair / car.repairs.size());
        }
    }
}
